### DISCUSSION TITLE
SPH 6000TL No Battery Stats

### DISCUSSION THREAD

**Discussion by cliver1956** [johanmeijer/grott#224] [Q&A]
Hi,
I have grott running and feeding data from my SPH6000 into influxdb 2.4.
My main use for it is to give me my solar generation and useage and battery stats.
I am a liitle confused as to what the metrics being stored are. Some are obvipous, others less so.
My main concern at the moment is that I don't seem to be getting any battery stats that I can identify.
Is there a document anywhere (I can't find one) that defines what each metric output by the Growatt is measuring  and how it refers to Solar generation, battery charge/discharge etc.?
The metrics I have available in infulxdb data explorer are:
![image](https://user-images.githubusercontent.com/52658257/200171570-a616fe20-6495-40fb-b329-0b3cf3096a16.png)

The Docker command I used to build the container is:
`docker run -d -p 5279:5279 --restart unless-stopped -e gnomqtt="True" -e gblockcmd="False" -e gpvoutput="False" -e ginflux="True" -e ginflux2="True" -e gifip="xx.xxx.xxx.xxx" -e gifport="8086" -e giftoken="xxxxxxxxxxx" -e giforg="********" -e gifbucket="*****" -e invtype="sph" --name grott ledidobe/grott`

The only thing I can think of is the 'invtype' environment variable. 
Have I got this right or is there something else I am missing?
Any help appreciated.

**Comment by githubDante** [comment]
Hi,

ginvtype=sph variable should help you to get additional data. The stats should be at the bottom. As for documentation, search for Growatt modbus RTU protocol. The registers for the SPH are described (don't expect much more) in the range 0-124, 1000-1124.

**Reply by cliver1956** [reply]
> Hi,
> 
> ginvtype=sph variable should help you to get additional data. The stats should be at the bottom. As for documentation, search for Growatt modbus RTU protocol. The registers for the SPH are described (don't expect much more) in the range 0-124, 1000-1124.

Thanks @githubDante , I'll search for that and have a read.
Every field that I have a measurement for is shown in the right hand column in my screenshot above. Non of them look like anything battery related to me?

**Comment by robupham** [comment]
I also have an SPH and am using the invtype=sph setting in my grott.ini

I see various stats related to the battery, including SOC (State of Charge) and various others that seem to be voltage related, although haven't dug into them all. See screenshot attached.

![image](https://user-images.githubusercontent.com/6196375/200373319-b9411b65-d7dc-4e7d-9a76-b32fd020db8c.png)

**Comment by cliver1956** [comment]
Strange, your schema is completely different to mine. I wonder why?

Regards
Clive
________________________________

**Reply by githubDante** [reply]
> Strange, your schema is completely different to mine. I wonder why? Regards Clive
> […](#)

Your Docker parameters are incorrect. See my reply (the environment variable is **ginvtype**). In your command you're using _-e invtype="sph"_

**Reply by cliver1956** [reply]
Brilliant, well spotted @githubDante . I don't have access at the moment but I'll amend and rebuild the container tomorrow.

**Reply by cliver1956** [reply]
Well, that seems to have worked. Supprising what happens when you get the syntax right :).
Thanks for the spot @githubDante

**Reply by cliver1956** [reply]
I can't mark your reply as the answer. How do we close this off? seems counter intuitive to mark my own comment as the answer?

**Comment by johanmeijer** [comment]
Can you send the Grott log at start up and the part where one of the records is being decrypted (including the layout defs used). Then we can see what goes wrong.

Just to be sure do you use the master branch or the beta (2.7.7)?

**Reply by cliver1956** [reply]
Thanks @johanmeijer as above, I don't have access at the moment. I'll try getting the environment variable right first thing and see how that goes.
I'm using the master brach.

**Reply by cliver1956** [reply]
All aseems to be working now I have the ginvtype set correctly. Thanks for your help @johanmeijer

**Reply by cliver1956** [reply]
Now I'm having trouble reconsiling the stats in my grottdb against the stats in server.growatt.com.
I am trying to monitor/report on.
1. Solar Panel output (2 strings)
2. Sytem production (Solar + Battery)
3. Battery SOC
4. Expoerted Power
5. Imported Power
Apart from SOC I can't figure which fields I need for each.
The info popups on my growatt dashboard are attached but most of them don't seem to match anything in grott?
Is there a list anywhere that describes exactly what each field is showing me or better still mapping what grott is storing to what growatt is showing?
![GrowattSystem](https://user-images.githubusercontent.com/52658257/203982555-f4187e3f-8f8d-4d76-a5b3-f3cc57c72dcc.jpg)
![GrowattEnergyTrend](https://user-images.githubusercontent.com/52658257/203982569-dc608c75-250f-4e47-b26d-665c7b04e737.jpg)

**Comment by johanmeijer** [comment]
Ok. I use a little bit off mixed terminologie in the Grott layouts (Just like grotwatt does online and in the documentation) Also not all fields are included. I will try to make a readable mapping. 

Calculated values are always difficult to trace back, but lets try. I do not believe that in the SPH records (at least the one I saw) the new version values are include. But to know for sure I have to see one of your data records ( The "- Grottserver - Original Data:" part),

**Reply by cliver1956** [reply]
> I have to see one of your data records ( The "- Grottserver - Original Data:" part),

@johanmeijer thanks for your reply. I'm really sorry but I'm not sure what you are asking me to show?
I'm a complete newbie with InfluxDB and flux. I have been accessing it mainly via indflubdb and grafana webui's mainly. 
I can access the command line terminal but not familiar with it.
Keen to learn but need a bit more guidance as to what you need me to show if you can please?
Thanks for the support so far BTW.

**Comment by johanmeijer** [comment]
I am looking for the Grott log with the original data part.

**Comment by johanmeijer** [comment]
Ok lets start.....

I published a documentation spreadsheet: 
https://github.com/johanmeijer/grott/blob/2.7-(Beta)/examples/Documentation/Growatt%20Layout%20definitions%20.xlsx

I used the original  Growatt documentation (Column A-F) and added the fields which I defined in the growatt layout (column G- J). 

For the SPH registers 1-124 and 1000-1124 are used and should be in the data record. 

I do not show everything in the Grott mqtt message (or influxdb). This layout is evolutionary created with data records from users and excel info for this records from growatt (not available anymore). 

Can you try to match things and see what you miss in the layout? It is always possible to change the layout definitions (only add for compatibility reasons) or create new layouts! Be aware some field names are mandatory for Grott processing.

**Reply by cliver1956** [reply]
Thanks @johanmeijer, I will get onto this but it may take a while as I am dealing with a family bereavement at the moment.

**Reply by cliver1956** [reply]
I got to have a short look at the spreadsheet today. 
I have extracted the data that I believe is appertaining to my SPH 6000 and added the fields available in my InfluxDB to the attached spreadsheet. I don't know if it is of any use to you or anyone else.
I then did a reconciliation between the Grott Field Names on your spreadsheet and the field names in my InfluxDB.
The ones Highlighted in Green have a one to one exact relationship between the two lists.
The ones in Orange appear more than once in your list.
The ones in Red do not appear in your list at all.
The ones in Green should be fine
The ones in Orange I guess will be fine as well assuming the muliple instances are all measuring the same thing?
As for the RED ones, I'm not sure any of them are important at the moment but I need to identify exactly what they are measuring to be sure.
On that note my next step is to identify exactly what each field is measuring and how it realates to the data I am seeing on the Growatt server so that I can replicate what I need and evetually use your grott server to cut China out of the loop altogether.

[grott_influxdb_data_fields.xlsx](https://github.com/johanmeijer/grott/files/10177502/grott_influxdb_data_fields.xlsx)

**Reply by muppet3000** [reply]
@johanmeijer - Sorry to revive this old thread, but I've recently started trying to use Grott and there are a few values that don't align to the values that I can get from the API.
I also have an SPH6000 inverter and I'm keen to help out and fix the incorrect values.
Can you please explain to me how the values in the spreadsheet that you've provided above [this one](https://github.com/johanmeijer/grott/blob/2.7-(Beta)/examples/Documentation/Growatt%20Layout%20definitions%20.xlsx) map onto the registers in the Growatt Modbus documentation.
e.g. There is the "First" column which is the register listed for the register, but then there's the "Grott Offset" which has a completely different value.
an example:
"pvstatus", has a register 'first' of 0, but then a 'grott offset' of 158. How do these items align?

The specific values that I believe are incorrect are `etogrid_tot` and `pvpowerout` but I'm struggling to investigate further without understanding how the registry values map onto the grott values.
If there's some documentation somewhere that explains this please point me towards it and I'll happily read it.

**Comment by johanmeijer** [comment]
Every group in the Modus register definition has an offset in the data record. 

For the SPH record reg 0 start at 158. Every register afters wordt in the first group start reg*4 so reg 124 (end of first group ) starts at 158+124*4. Some values like input power has a register for the High and Low part. So input_power has a total length of 8. 

The second group for the SPH (1000-1124) starts at offset 666. To calculate the offset of a register in this group you can use formula: 666+(reg-1000)*4.  For the last register in this group 666+(1124-1000)*4=1162, 

Be aware some values on the Growatt website/App does not have direct relation with a register. Some values are calculated. 
And some values points to same registers (or only filled if a register is positive or negative). 
The only way to detect how that works is to do some reverse engineering.
